### PR TITLE
BUG: cast to correct dtype in Index.drop()

### LINKED ISSUE #18304: pd.Index([ ('b', 'c'), 'a']).drop(['a', ('b', 'c')]) raises ValueError
#### Code Sample, a copy-pastable example if possible

```bash
pietro@debiousci:~$ PYTHONHASHSEED=5 python3 -c "import pandas as pd; s1 = pd.Series([0,1], name='a'); s2 = pd.Series([2,3], name=('b', 'c')); print(pd.crosstab(s1, s2))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/pietro/nobackup/repo/pandas/pandas/core/reshape/pivot.py", line 466, in crosstab
    dropna=dropna, **kwargs)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/frame.py", line 4462, in pivot_table
    margins_name=margins_name)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/reshape/pivot.py", line 82, in pivot_table
    agged = grouped.agg(aggfunc)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/groupby.py", line 4191, in aggregate
    return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/groupby.py", line 3632, in aggregate
    return self._python_agg_general(arg, *args, **kwargs)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/groupby.py", line 873, in _python_agg_general
    return self._wrap_aggregated_output(output)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/groupby.py", line 4254, in _wrap_aggregated_output
    agg_labels = self._obj_with_exclusions._get_axis(agg_axis)
  File "pandas/_libs/properties.pyx", line 39, in pandas._libs.properties.cache_readonly.__get__ (pandas/_libs/properties.c:1604)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/base.py", line 235, in _obj_with_exclusions
    return self.obj.drop(self.exclusions, axis=1)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/generic.py", line 2517, in drop
    obj = obj._drop_axis(labels, axis, level=level, errors=errors)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/generic.py", line 2549, in _drop_axis
    new_axis = axis.drop(labels, errors=errors)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/indexes/base.py", line 3750, in drop
    labels = _index_labels_to_array(labels)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/common.py", line 417, in _index_labels_to_array
    labels = _asarray_tuplesafe(labels)
  File "/home/pietro/nobackup/repo/pandas/pandas/core/common.py", line 386, in _asarray_tuplesafe
    result = np.asarray(values, dtype=dtype)
  File "/home/pietro/.local/lib/python3.5/site-packages/numpy/core/numeric.py", line 531, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: setting an array element with a sequence
```
Compare to:

```
pietro@debiousci:~$ PYTHONHASHSEED=6 python3 -c "import pandas as pd; s1 = pd.Series([0,1], name='a'); s2 = pd.Series([2,3], name=('b', 'c')); print(pd.crosstab(s1, s2))"
('b', 'c')  2  3
a               
0           1  0
1           0  1
```

#### Problem description

The above happens (pseudo-)randomly with python 3 and, it seems, always with python 2.

#### Expected Output

The case `` PYTHONHASHSEED=6``.

#### Output of ``pd.show_versions()``

<details>

In [2]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.22.0.dev0+131.g63e8527d3
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.7.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: None
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #18304
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can you add the test from the original issue. (crosstab), lgtm. otherwise.

File: pandas/core/common.py [hunk 1]
```diff
@@ -404,7 +404,7 @@ def _asarray_tuplesafe(values, dtype=None):
     return result
 
 
-def _index_labels_to_array(labels):
+def _index_labels_to_array(labels, dtype=None):
```
Comment (jreback) [inline pandas/core/common.py]: doc-string here would be nice; in theory some unit tests if you can.

Comment (toobaz) [reply pandas/core/common.py]: docstring added; this is tested in my new test and, in old tests, whenever ``Index(dtype='object').drop()`` is called.

Comment (jreback) [reply pandas/core/common.py]: I am asking for some unit-tests for this routine itself. all of these helper routines should ideally have tests.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/common.py (hunk 1 above)
Comment (toobaz) [reply pandas/core/common.py]: Uhm... unless there are specific coverage (lines of code) issues, I personally don't see this as an efficient use of my time, considering how many and how frequently changed are pandas internal routines.

**Comment by jreback** [comment]
can you rebase

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @toobaz

**Pull request merged**